service/storage_service.cc

```diff
@@ -2257,6 +2257,143 @@ future<> storage_service::removenode(sstring host_id_string) {
+                    });
+                }).get();
+
+                // Step 4: Announce the node has left
+                std::unordered_set<token> tmp(tokens.begin(), tokens.end());
+                ss.excise(std::move(tmp), endpoint);
+                ss._gossiper.advertise_token_removed(endpoint, host_id).get();
+                slogger.info("removenode[{}]: Finshed removenode operation, removing node={}, nodes to sync data={}, nodes to skip sync data={}", uuid, endpoint, nodes, skip_sync_on_nodes);
+            } catch (...) {
+                // we need to revert the effect of prepare verb the removenode ops is failed
+                req.cmd = removenode_cmd::abort;
+                parallel_for_each(nodes, [&ss, &req, &nodes_unknown_verb, &nodes_down, uuid] (const gms::inet_address& node) {
+                    if (nodes_unknown_verb.contains(node) || nodes_down.contains(node)) {
+                        // No need to revert previous prepare cmd for those who do not apply prepare cmd.
+                        return make_ready_future<>();
+                    }
+                    return ss._messaging.local().send_removenode(netw::msg_addr(node), req).then([uuid, node] (removenode_response resp) {
+                        slogger.info("removenode[{}]: Got abort response from node={}", uuid, node);
+                    });
+                }).get();
+                slogger.info("removenode[{}]: Aborted removenode operation, removing node={}, nodes to sync data={}, nodes to skip sync data={}", uuid, endpoint, nodes, skip_sync_on_nodes);
+                throw;
+            }
+
+        });
+    });
+}
+#endif
+
+future<removenode_response> storage_service::removenode_handler(gms::inet_address coordinator, removenode_request req) {
+    return get_storage_service().invoke_on(0, [coordinator, req = std::move(req)] (auto& ss) mutable {
+        return seastar::async([&ss, coordinator, req = std::move(req)] () mutable {
+            auto ops_uuid = req.req_id;
+            if (req.leaving_nodes.size() > 1) {
+                auto msg = format("removenode[{}]: Could not removenode more than one node at a time: leaving_nodes={}", req.req_id, req.leaving_nodes);
+                slogger.warn("{}", msg);
+                throw std::runtime_error(msg);
+            }
+            if (req.cmd == removenode_cmd::prepare) {
+                auto tmlock = std::make_unique<token_metadata_lock>(ss.get_token_metadata_lock().get0());
+                auto tmptr = ss.get_mutable_token_metadata_ptr().get0();
+                for (auto& node : req.leaving_nodes) {
+                    slogger.info("removenode[{}]: Added node={} as leaving node, coordinator={}", req.req_id, node, coordinator);
+                    tmptr->add_leaving_endpoint(node);
+                }
+                ss.update_pending_ranges(tmptr, format("removenode {}", req.leaving_nodes)).get();
+                ss.replicate_to_all_cores(std::move(tmptr)).get();
```

Comment:
better to use mutate_token_metadata(), less things that can go wrong.

Reply:
ok. I will switch to the new helper.